hmwk/hmwk3.py

```diff
@@ -0,0 +1,97 @@
+import numpy as np
+from tasks.week6.mangle import create_mangle_file
+from tasks.week6.spherical_caps import  ra_vect_array, dec_vect_array, cap
+from astropy.coordinates import Angle, SkyCoord
+import pymangle
+import astropy.units as u
+from hmwk.hmwk2 import rect_area
+from astropy.table import Table
+import matplotlib.pyplot as plt
+
+"""
+This script constructs mangle masks for a rectangular area and several plates mostly
+within the rectangle, and finds the area where the plates and rectangle overlap.
+It then plots quasars, indicating those quasars that fall within the area of the plates.
+"""
+    
+# KFH Use functions from week6 to generate caps
+# KFH Generate caps for rectangular survey area
+ra_min_cap = ra_vect_array('10h15m00s')
+ra_max_cap = ra_vect_array('11h15m00s', neg=True)
+dec_min_cap = dec_vect_array('+30d00m00s')
+dec_max_cap = dec_vect_array('+40d00m00s', neg=True)
+
+# KFH Generate caps for plates
+plate_cap1 = cap(155*u.deg, 34*u.deg, 2)
+plate_cap2 = cap(159*u.deg, 36*u.deg, 2)
+plate_cap3 = cap(163*u.deg, 34*u.deg, 2)
+plate_cap4 = cap(167*u.deg, 36*u.deg, 2)
+
+# KFH Create mangle of rectangle only
+create_mangle_file(np.array([[ra_min_cap, ra_max_cap, dec_min_cap, dec_max_cap]]), 'survey_area.ply')
+
+# KFH Create array of polys that each include the survey rectangle and one plate
+poly_array = np.asarray([[plate_cap1, ra_min_cap, ra_max_cap, dec_min_cap, dec_max_cap],
+                    [plate_cap2, ra_min_cap, ra_max_cap, dec_min_cap, dec_max_cap],
+                    [plate_cap3, ra_min_cap, ra_max_cap, dec_min_cap, dec_max_cap],
+                    [plate_cap4, ra_min_cap, ra_max_cap, dec_min_cap, dec_max_cap]])
+create_mangle_file(poly_array, "plate_survey_overlap.ply", weight=[1,1,1,1], area=[0,0,0,0])
+
+# KFH Create mangle objects
+rect = pymangle.Mangle('survey_area.ply')
+m = pymangle.Mangle('plate_survey_overlap.ply')
+
+# KFH Randomly generate equal-distributed points in rectangle
+points = 100000
+ra, dec = rect.genrand(points)
```

Comment:
It was smart to generate the points just in the rectangle to perform the "Monte Carlo" integration. This sped your code up considerably compared to generating points over the entire sphere.